File: src/test/java/com/telegram/codex/conversation/application/reply/ReplyGenerationServiceTest.java
package com.telegram.codex.conversation.application.reply;

import com.telegram.codex.conversation.application.ProcessedUpdateService;
import com.telegram.codex.conversation.application.gateway.ReplyGenerationGateway;
import com.telegram.codex.conversation.application.session.SessionService;
import com.telegram.codex.conversation.domain.memory.ChatMemoryRecord;
import com.telegram.codex.conversation.domain.session.ChatSessionRecord;
import com.telegram.codex.conversation.infrastructure.memory.ChatMemoryRepository;
import com.telegram.codex.conversation.infrastructure.memory.CodexMemoryClient;
import com.telegram.codex.conversation.infrastructure.session.ChatSessionRepository;
import com.telegram.codex.integration.telegram.application.port.out.TelegramGateway;
import com.telegram.codex.integration.telegram.domain.InboundMessage;
import org.junit.jupiter.api.Test;
import org.mockito.Mockito;

import java.util.List;
import java.util.Optional;
import java.util.function.Supplier;

import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.anyList;
import static org.mockito.ArgumentMatchers.anyString;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

class ReplyGenerationServiceTest {

    @Test
    void handleUsesStoredConversationState() {
        ReplyGenerationGateway cliClient = Mockito.mock(ReplyGenerationGateway.class);
        CodexMemoryClient memoryClient = Mockito.mock(CodexMemoryClient.class);
        ProcessedUpdateService processedUpdateService = Mockito.mock(ProcessedUpdateService.class);
        SessionService sessionService = Mockito.mock(SessionService.class);
        TelegramGateway telegramGateway = Mockito.mock(TelegramGateway.class);
        AttachmentDownloader attachmentDownloader = Mockito.mock(AttachmentDownloader.class);
        when(cliClient.generateReply(anyString(), any(), anyList(), any(), any())).thenReturn(new ReplyResult("next-state", List.of("a", "b", "c"), "reply"));
        ChatSessionRepository sessionRepository = mockSessionRepository(Optional.of(new ChatSessionRecord("3", "[{\"role\":\"user\",\"content\":\"hi\"}]", System.currentTimeMillis())));
        ChatMemoryRepository memoryRepository = Mockito.mock(ChatMemoryRepository.class);
        when(memoryRepository.find("3")).thenReturn(Optional.of(new ChatMemoryRecord("3", "記憶", System.currentTimeMillis())));
        when(memoryClient.merge("記憶", "你好", "reply")).thenReturn("記憶");
        when(attachmentDownloader.downloadImages(List.of())).thenReturn(List.of());
        when(telegramGateway.withTypingStatus(eq("3"), any())).thenAnswer(invocation -> invocation.getArgument(1, Supplier.class).get());

        ReplyGenerationService service = new ReplyGenerationService(
            cliClient,
            sessionRepository,
            memoryRepository,
            memoryClient,
            processedUpdateService,
            sessionService,
            telegramGateway,
            attachmentDownloader
        );

        service.handle(buildMessage());

        verify(processedUpdateService).pruneIfNeeded();
        verify(processedUpdateService).savePendingReply(99L, "3", 10L, new ReplyResult("next-state", List.of("a", "b", "c"), "reply"));
        verify(telegramGateway).sendMessage("3", "reply", List.of("a", "b", "c"), false);
        verify(sessionService).persistConversationState("3", "next-state");
        verify(processedUpdateService).markProcessed(99L, "3", 10L);
        verify(attachmentDownloader).downloadImages(List.of());
        verify(attachmentDownloader).cleanup(List.of());
    }

    @Test
    void handleSkipsLongTermMemoryRefreshForBlankUserText() {
        ReplyGenerationGateway cliClient = Mockito.mock(ReplyGenerationGateway.class);
        ProcessedUpdateService processedUpdateService = Mockito.mock(ProcessedUpdateService.class);
        SessionService sessionService = Mockito.mock(SessionService.class);
        TelegramGateway telegramGateway = Mockito.mock(TelegramGateway.class);
        AttachmentDownloader attachmentDownloader = Mockito.mock(AttachmentDownloader.class);
        CodexMemoryClient memoryClient = Mockito.mock(CodexMemoryClient.class);
        ChatMemoryRepository memoryRepository = Mockito.mock(ChatMemoryRepository.class);
        when(cliClient.generateReply(any(), any(), any(), any(), any())).thenReturn(new ReplyResult("next-state", List.of(), "reply"));
        when(memoryRepository.find("3")).thenReturn(Optional.empty());
        when(attachmentDownloader.downloadImages(List.of())).thenReturn(List.of());
        when(telegramGateway.withTypingStatus(eq("3"), any())).thenAnswer(invocation -> invocation.getArgument(1, Supplier.class).get());
        ChatSessionRepository sessionRepository = mockSessionRepository(Optional.empty());

        ReplyGenerationService service = new ReplyGenerationService(
            cliClient,
            sessionRepository,
            memoryRepository,
            memoryClient,
            processedUpdateService,
            sessionService,
            telegramGateway,
            attachmentDownloader
        );

        service.handle(new InboundMessage("3", List.of(), null, 10, List.of(), List.of(), null, null, "   ", "5", 99));

        verify(memoryClient, never()).merge(anyString(), anyString(), anyString());
        verify(memoryRepository, never()).persist(anyString(), anyString());
    }

    private ChatSessionRepository mockSessionRepository(Optional<ChatSessionRecord> record) {
        ChatSessionRepository repository = Mockito.mock(ChatSessionRepository.class);
        when(repository.findActive("3")).thenReturn(record);
        return repository;
    }

    private InboundMessage buildMessage() {
        return new InboundMessage("3", List.of(), null, 10, List.of(), List.of(), null, null, "你好", "5", 99);
    }
}

File: src/test/kotlin/com/telegram/codex/conversation/application/ReplyGenerationServiceTest.kt
package com.telegram.codex.conversation.application

import com.telegram.codex.conversation.domain.ChatMemoryRecord
import com.telegram.codex.conversation.domain.ChatSessionRecord
import com.telegram.codex.conversation.infrastructure.ChatMemoryRepository
import com.telegram.codex.conversation.infrastructure.ChatSessionRepository
import com.telegram.codex.conversation.infrastructure.CodexMemoryClient
import com.telegram.codex.integration.telegram.application.TelegramGateway
import com.telegram.codex.integration.telegram.domain.InboundMessage
import com.telegram.codex.integration.telegram.domain.TelegramBotCommand
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Test
import org.mockito.Mockito
import org.mockito.Mockito.verify
import java.nio.file.Path

class ReplyGenerationServiceTest {
    @Test
    fun handleUsesStoredConversationState() {
        val cliClient = Mockito.mock(ReplyGenerationGateway::class.java)
        val memoryClient = Mockito.mock(CodexMemoryClient::class.java)
        val processedUpdateService = Mockito.mock(ProcessedUpdateService::class.java)
        val sessionService = Mockito.mock(SessionService::class.java)
        val telegramGateway = FakeTelegramGateway()
        val attachmentDownloader = Mockito.mock(AttachmentDownloader::class.java)
        Mockito.`when`(cliClient.generateReply(Mockito.anyString(), Mockito.any(), Mockito.anyList(), Mockito.any(), Mockito.any()))
            .thenReturn(ReplyResult("next-state", listOf("a", "b", "c"), "reply"))
        val sessionRepository = Mockito.mock(ChatSessionRepository::class.java)
        Mockito.`when`(sessionRepository.findActive("3")).thenReturn(ChatSessionRecord("3", "state", System.currentTimeMillis()))
        val memoryRepository = Mockito.mock(ChatMemoryRepository::class.java)
        Mockito.`when`(memoryRepository.find("3")).thenReturn(ChatMemoryRecord("3", "記憶", System.currentTimeMillis()))
        Mockito.`when`(memoryClient.merge("記憶", "你好", "reply")).thenReturn("記憶")
        Mockito.`when`(attachmentDownloader.downloadImages(emptyList())).thenReturn(emptyList())

        val service = ReplyGenerationService(cliClient, sessionRepository, memoryRepository, memoryClient, processedUpdateService, sessionService, telegramGateway, attachmentDownloader)
        val message = InboundMessage("3", emptyList(), null, 10, emptyList(), emptyList(), null, "你好", "5", 99)

        service.handle(message)

        verify(processedUpdateService).pruneIfNeeded()
        verify(processedUpdateService).savePendingReply(99L, "3", 10L, ReplyResult("next-state", listOf("a", "b", "c"), "reply"))
        assertEquals(FakeTelegramGateway.SentMessage("3", "reply", listOf("a", "b", "c"), false), telegramGateway.sentMessage)
        verify(sessionService).persistConversationState("3", "next-state")
        verify(processedUpdateService).markProcessed(99L, "3", 10L)
    }

    private class FakeTelegramGateway : TelegramGateway {
        var sentMessage: SentMessage? = null

        override fun downloadFileToTemp(fileId: String): Path = throw UnsupportedOperationException()
        override fun sendMessage(chatId: String, text: String?, suggestedReplies: List<String>, removeKeyboard: Boolean) {
            sentMessage = SentMessage(chatId, text, suggestedReplies, removeKeyboard)
        }
        override fun <T> withTypingStatus(chatId: String, action: () -> T): T = action()
        override fun setWebhook(url: String, secretToken: String) = Unit
        override fun setMyCommands(commands: List<TelegramBotCommand>) = Unit

        data class SentMessage(val chatId: String, val text: String?, val suggestedReplies: List<String>, val removeKeyboard: Boolean)
    }
}
